nncf/experimental/post_training/graph/model_transformer.py

```diff
@@ -0,0 +1,96 @@
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+from abc import ABC
+from abc import abstractmethod
+from typing import List
+from typing import TypeVar
+
+from nncf.common.graph.model_transformer import ModelTransformer
+from nncf.common.graph.transformations.commands import TransformationCommand
+from nncf.common.graph.transformations.layout import TransformationLayout
+from nncf.experimental.post_training.statistics.statistic_point import StatisticPointsContainer
+
+ModelType = TypeVar('ModelType')
+
+
+# pylint: disable=no-member
+class StaticModelTransformerBase(ModelTransformer, ABC):
+
+    def __init__(self, model: ModelType):
+        super().__init__(model)
+        self._transformation_layout = None
+        self._transformations_list = []
+
+    def set_model(self, model: ModelType) -> None:
+        """
+        Sets model
+
+        :param model: input model
+        """
+        self._model = model
+
+    def prepare_model_for_statistics_collection(self, statistic_points: StatisticPointsContainer) -> ModelType:
+        """
+        Prepares model for statics collection by adding external outputs
+
+        :param statistic_points: StatisticPointsContainer
+        :return: model after transformations
+        """
+        transformation_layout = self._get_transformation_layout_extra_outputs(statistic_points)
+        return self.transform(transformation_layout)
+
+    def transform(self, transformation_layout: TransformationLayout) -> ModelType:
+        """
+        Applies transformation layout on the model
+
+        :param transformation_layout: TransformationLayout
+        :return: model after transformations
+        """
+        self._apply_transformations(transformation_layout.transformations)
+        return self._model
+
+    @abstractmethod
+    def _apply_transformations(self, transformations: List[TransformationCommand]) -> None:
+        """
+        Applies transformations by type-callback on the model
+
+        :param transformations: lisf of the TransformationCommand transformations
+        """
+
+    @abstractmethod
+    def _get_transformation_layout_extra_outputs(
+            self,
+            statistic_points: StatisticPointsContainer) -> TransformationLayout:
+        """
+        Collects transformations layout by statistic_points
+
+        :param statistic_points: StatisticPointsContainer
+        :return: transformation_layout
+        """
+
+    @abstractmethod
+    def _apply_quantizer_insertion_transformations(self, transformations: List[TransformationCommand]) -> None:
```

Comment:
This does not look good really if we need to have ```_apply-*``` method for each type of transformation. Can we have one ```_apply`` method that will execute a set of transformations, let's say, from a current buffer of transformations?

Reply:
This is a good suggestion. But how we will ask the developer to implement all required transformations during introducing the new backend without defining all transformations as `@abstractmethod`?

Reply:
Could you clarify why you need _apply-* methods as abstract methods? It looks like you should implement only one method of ModelTransformer API. This is ``` def transform(self, transformation_layout: TransformationLayout) -> ModelType:```

Reply:
@alexsu52, this is a good question, and a partial answer is above.
The main reason to make these methods abstract is that the developer should realize this method in the other backend-specific `ModelTransformer`. If there are any suggestions on how to make these methods required for re-implementation in another instance - I'm open to the suggestions :)

Reply:
Thanks for the quick reply! Another option is implementation of support for defined list of transformation commands. It looks like ```InsertionComand``` and ```OutputInsertionCommand``` should be supported.

Reply:
@alexsu52, making these transformation commands as required doesn't guarantee that corresponding transformations will be realized in the new backend-specific `ModelTransformer`, right?
Also, there was another solution for this moment. Please, see here - https://github.com/openvinotoolkit/nncf/pull/1264/commits/aa2d6efa282eb0a2c1caff2eeeef0c8ef3c6762b#diff-8e587534fd229e04f68ba4153308f02ad4eba805a228b20756e6e723622db6c2R77
But reviewers said this solution is too complicated and it's better to have multiple required transformations and `if` conditions.